### PR TITLE
Fix ilm race condition

### PR DESCRIPTION
**Pull request by sscimff**
Fixes race condition in ILM Shrink action (#109206). Ensures ILM execution state is copied before execution, improving stability.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/ClusterStateObserver.java [hunk 1]
```diff
@@ -30,6 +30,7 @@
  */
 public class ClusterStateObserver {
 
+    public static org.elasticsearch.xpack.core.ml.MlMetadata.Builder ChangePredicate;
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/ClusterStateObserver.java:33]: Can we revert this?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkStep.java [hunk 2]
```diff
@@ -98,10 +98,37 @@ public void performAction(
...
+                    l.onFailure(new IllegalStateException("Timed out waiting for shrunk index to be allocated"));
+                }
+            };
+
+            observer.waitForNextChange(clusterStateListener);
+        }));
+    }
+    private boolean shrunkenIndexAllocated(ClusterState clusterState, String shrunkenIndexName) {
+        IndexMetadata indexMetadata = clusterState.metadata().index(shrunkenIndexName);
+        return indexMetadata != null && indexMetadata.getState() == IndexMetadata.State.OPEN;
```
Comment (henningandersen) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkStep.java:128]: This does not ensure that the index is allocated, only that it exists and is open.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkStep.java [hunk 3]
```diff
@@ -98,10 +98,37 @@ public void performAction(
         // Hard coding this to true as the resize request was executed and the corresponding cluster change was committed, so the
         // eventual retry will not be able to succeed anymore (shrunk index was created already)
         // The next step in the ShrinkAction will wait for the shrunk index to be created and for the shards to be allocated.
-        getClient().admin().indices().resizeIndex(resizeRequest, listener.delegateFailureAndWrap((l, response) -> l.onResponse(null)));
-
+        getClient().admin().indices().resizeIndex(resizeRequest, listener.delegateFailure((l, response) -> {
+            ClusterStateObserver.Listener clusterStateListener = new ClusterStateObserver.Listener() {
+                @Override
+                public void onNewClusterState(ClusterState state) {
```
Comment (henningandersen) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkStep.java:104]: This is not the right way to use an observer, instead you should pass the condition as a predicate to `waitForNextChange`.

**Review by henningandersen — CHANGES_REQUESTED** [review]
> Ensures ILM execution state is copied before execution, improving stability

I am not sure I understand this, can you elaborate? This does not look like it copies execution state, rather it waits for an always true condition AFAICS.

We also need a test demonstrating the issue, will you add that too?

**Comment by nielsbauman** [comment]
Hi @sscimff, thank you for your interest in Elasticsearch and for taking the time to open a PR! I see you haven't addressed @henningandersen's comments yet. Are you still planning to work on this PR? The reason I'm asking is that we've discussed this issue (https://github.com/elastic/elasticsearch/issues/109206) internally, and we're inclined to go with a different approach from the one you've taken here.

If you're still interested in fixing this issue, I'm happy to explain our proposed solution to you. Otherwise, I'll close your PR and open one of my own.

**Comment by sscimff** [comment]
Hi @nielsbauman thank you for the message. If you have another approach and decide to implement it. Please go ahead and close this PR.

**Comment by nielsbauman** [comment]
We've decided internally on a different approach for this issue (see https://github.com/elastic/elasticsearch/issues/109206#issuecomment-2310376610). We'll be picking this up internally as well.

I'm going to close this PR, thanks nonetheless for your efforts @sscimff!

**Pull request closed without merging**